Source/WebGPU/WebGPU/Queue.mm

```diff
@@ -732,6 +732,139 @@ static void invalidateCommandBuffers(Vector<Ref<WebGPU::CommandBuffer>>&& comman
+        return;
+
+    MTLTextureDescriptor* sourceTextureDescriptor = [MTLTextureDescriptor new];
+    sourceTextureDescriptor.width = IOSurfaceGetWidth(sourceIOSurface);
+    sourceTextureDescriptor.height = IOSurfaceGetHeight(sourceIOSurface);
+    sourceTextureDescriptor.usage = MTLTextureUsageShaderRead;
+    sourceTextureDescriptor.textureType = MTLTextureType2D;
+    sourceTextureDescriptor.pixelFormat = MTLPixelFormatBGRA8Unorm;
+
+    Ref destinationTexture = fromAPI(destination.texture);
+    id<MTLTexture> destinationMTLTexture = destinationTexture->texture();
+    destinationTexture->setPreviouslyCleared();
+
+    float destinationWidth = static_cast<float>(destinationMTLTexture.width);
+    float destinationHeight = static_cast<float>(destinationMTLTexture.height);
+
+    float xOffset = destination.origin.x / destinationWidth;
+    float yOffset = destination.origin.y / destinationHeight;
+    float xSize = width / destinationWidth;
+    float ySize = height / destinationHeight;
+
+    RELEASE_ASSERT(!flipY);
+    RELEASE_ASSERT(xSize + xOffset <= 1.f);
+    RELEASE_ASSERT(ySize + yOffset <= 1.f);
+    float uvs[] = {
+        originX / static_cast<float>(sourceTextureDescriptor.width), originY / static_cast<float>(sourceTextureDescriptor.height),
+        0.f   + xOffset,    0.f   + yOffset,
+        xSize + xOffset,    0.f   + yOffset,
+        0.f   + xOffset,    ySize + yOffset,
+
+        xSize + xOffset,    0.f   + yOffset,
+        xSize + xOffset,    ySize + yOffset,
+        0.f   + xOffset,    ySize + yOffset
+    };
+    id<MTLTexture> sourceTexture = device->newTextureWithDescriptor(sourceTextureDescriptor, sourceIOSurface);
+
+    id<MTLRenderPipelineState> pso = copyPSO(destinationMTLTexture);
+    if (!pso)
+        return;
+
+    Ref deviceQueue = device->getQueue();
+
+    MTLRenderPassDescriptor* renderPassDescriptor = [MTLRenderPassDescriptor new];
+    renderPassDescriptor.colorAttachments[0].storeAction = MTLStoreActionStore;
+    renderPassDescriptor.colorAttachments[0].loadAction = MTLLoadActionDontCare;
+    renderPassDescriptor.colorAttachments[0].texture = destinationMTLTexture;
+    renderPassDescriptor.colorAttachments[0].slice = destination.origin.z;
+
+    MTLCommandBufferDescriptor *descriptor = [MTLCommandBufferDescriptor new];
+    id<MTLCommandBuffer> commandBuffer = deviceQueue->commandBufferWithDescriptor(descriptor);
+
+    id<MTLRenderCommandEncoder> renderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:renderPassDescriptor];
+    deviceQueue->setEncoderForBuffer(commandBuffer, renderEncoder);
+    [renderEncoder setRenderPipelineState:pso];
+
+    [renderEncoder setVertexBytes:uvs length:sizeof(uvs) atIndex:0];
+    [renderEncoder setFragmentTexture:sourceTexture atIndex:0];
+    [renderEncoder drawPrimitives:MTLPrimitiveTypeTriangle vertexStart:0 vertexCount:6];
+    deviceQueue->endEncoding(renderEncoder, commandBuffer);
+    deviceQueue->commitMTLCommandBuffer(commandBuffer);
```

Comment:
Committing the command buffer here is likely not needed, we could defer it in the case of multiple uploads in order to coalesce them.